### PR TITLE
Multitag support (#202)

### PR DESCRIPTION
**Pull request by ajurasz**
Hi, here is my proposal for handling multiple tags in `hazelcast-aws`. If it comes to parsing `tag-key` and `tag-value` I did it as described [here](https://github.com/hazelcast/hazelcast-aws/issues/202#issuecomment-766418503) (I kept the possibility to filter instances using tag value(s) as this can still be a valid use case).

I did some testing on AWS and from my perspective, it behaves as expected. Setup was the following:

```
172.31.23.8 - testing instance
172.31.22.87  - tags = env:DEV, service:auth
172.31.16.117 - tags = env:DEV, service:pricing
172.31.16.162 - tags = env:DEV, service:auth
```

## Test case 1:

Configuration:

```yml
hazelcast:
  network:
    join:
      multicast:
        enabled: false
      aws:
        enabled: true
        iam-role: hazelcast_discovery_role
        hz-port: 5701
        region: us-east-1
        tag-key: env
        tag-value: DEV
    rest-api:
      enabled: true
```

Log output:

```
Jan 31, 2021 4:34:27 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [172.31.23.8]:5701 [dev] [4.1.1] Initialized new cluster connection between /172.31.23.8:50501 and /172.31.16.162:5701
Jan 31, 2021 4:34:27 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [172.31.23.8]:5701 [dev] [4.1.1] Initialized new cluster connection between /172.31.23.8:60527 and /172.31.16.117:5701
Jan 31, 2021 4:34:27 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [172.31.23.8]:5701 [dev] [4.1.1] Initialized new cluster connection between /172.31.23.8:35567 and /172.31.22.87:5701
```

Result:
Connections to all instances with tag `env:DEV` were initialized.

## Test case 2:

Configuration:

```yml
hazelcast:
  network:
    join:
      multicast:
        enabled: false
      aws:
        enabled: true
        iam-role: hazelcast_discovery_role
        hz-port: 5701
        region: us-east-1
        tag-key: env,service
        tag-value: DEV,auth
    rest-api:
      enabled: true
```

Log output:

```
Jan 31, 2021 4:37:39 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [172.31.23.8]:5701 [dev] [4.1.1] Initialized new cluster connection between /172.31.23.8:54671 and /172.31.16.162:5701
Jan 31, 2021 4:37:39 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [172.31.23.8]:5701 [dev] [4.1.1] Initialized new cluster connection between /172.31.23.8:48203 and /172.31.22.87:5701
```

Result:
Connections to all instances with tags `env:DEV` AND `service:auth` were initialized.

### CODE REVIEW

**Comment by ajurasz** [comment]
Since support for wildcard usage was trivial, I have included it into this PR.

File: src/main/java/com/hazelcast/aws/Tags.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        return new Tags(tags);
+    }
+
+    private static Collection<String> splitValue(String value) {
+        return isEmpty(value) ? Collections.emptyList() : Arrays.asList(value.split(SEPARATOR));
+    }
+
+    @Override
+    public Iterator<Tag> iterator() {
```
Comment (alparslanavci) [inline src/main/java/com/hazelcast/aws/Tags.java:74]: Could you reorganize methods down to follow [the Stepdown Rule](https://dzone.com/articles/the-stepdown-rule)?

File: src/test/java/com/hazelcast/aws/TagsTest.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        // and
+        tag = iterator.next();
+        assertNull(tag.getKey());
+        assertEquals("ValueB", tag.getValue());
+    }
+
+    @Test
+    public void noTags() {
+        // when
```
Comment (alparslanavci) [inline src/test/java/com/hazelcast/aws/TagsTest.java:113]: Nit: it would be great if you can define given/when/then steps for this test too.

**Review by alparslanavci — APPROVED** [review]
@ajurasz thank you for the PR, just added two small comments. 

LGTM, but I strongly suggest to wait for @leszko's review. He knows the internals better, maybe he might ask some changes in design.

File: src/test/java/com/hazelcast/aws/TagsTest.java (hunk 2 above)
Comment (ajurasz) [reply src/test/java/com/hazelcast/aws/TagsTest.java:117]: Since this test was covering to cases I have divided it into two and added `given` block for each.

File: src/main/java/com/hazelcast/aws/Tags.java (hunk 1 above)
Comment (ajurasz) [reply src/main/java/com/hazelcast/aws/Tags.java:74]: Went through the article but still not sure if what I have done is what is expected :)

File: src/main/java/com/hazelcast/aws/AwsConfig.java [hunk 3]
```diff
@@ -30,8 +30,7 @@
     private final String region;
     private final String hostHeader;
     private final String securityGroupName;
-    private final String tagKey;
-    private final String tagValue;
+    private final Tags tags;
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/AwsConfig.java:33]: I think that having a separate class `Tags` does not simplify anything here. I guess having here `List<Tag>` would be cleaner. And the class `Tag` can be either a separate class or an inner class of `AwsConfig`.

File: src/main/java/com/hazelcast/aws/AwsRequestUtils.java [hunk 4]
```diff
@@ -83,7 +83,9 @@ private static void addComponents(List<String> components, Map<String, String> a
     private static String urlEncode(String string) {
         String encoded;
         try {
-            encoded = URLEncoder.encode(string, "UTF-8").replace("+", "%20");
+            encoded = URLEncoder.encode(string, "UTF-8")
+                    .replace("+", "%20")
+                    .replace("*", "%2A");
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/AwsRequestUtils.java:88]: I wonder, why this change is needed?

File: src/main/java/com/hazelcast/aws/TagFilters.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+package com.hazelcast.aws;
+
+import com.hazelcast.aws.Tags.Tag;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+import static com.hazelcast.aws.StringUtils.isNotEmpty;
+
+final class TagFilters {
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/TagFilters.java:25]: This is just a util class with one method. I think you can move (and rename) this method to `AwsEc2Api`, since it's only used there. Or you can move it to the `Filter` class if you prefer.

**Review by leszko — CHANGES_REQUESTED** [review]
Thanks a lot for the PR @ajurasz . Sorry for the late review, but I was OOO for 2 weeks. Added my comments. Apart from that, really great work and I'm happy to see this feature.

File: src/main/java/com/hazelcast/aws/AwsRequestUtils.java (hunk 4 above)
Comment (ajurasz) [reply src/main/java/com/hazelcast/aws/AwsRequestUtils.java:88]: In the [documentation for signing AWS requests](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html), they write that only unreserved characters defined by RFC 3986 should not be URI-encoded. `*` is not part of these characters so it should be "Percent-encode" as another part of the documentation specifies.

File: README.md [hunk 6]
```diff
@@ -100,7 +100,7 @@ The following properties can be configured (all are optional).
 * `region`: region where Hazelcast members are running; default is the current region
 * `host-header`: `ec2`, `ecs`, or the URL of a EC2/ECS API endpoint; automatically detected by default
 * `security-group-name`: filter to look only for EC2 instances with the given security group
-* `tag-key`, `tag-value`: filter to look only for EC2 Instances with the given `tag-key`/`tag-value`
+* `tag-key`, `tag-value`: filter to look only for EC2 Instances with the given `tag-key`/`tag-value`; multi values supported if comma separated (e.g. `KeyA,KeyB`)
```
[the remark this answers is not in the document]
Comment (ajurasz) [reply README.md]: This is true. I have put additional information about, what kind of condition these comma-separated values build.

File: src/main/java/com/hazelcast/aws/TagFilters.java (hunk 5 above)
Comment (ajurasz) [reply src/main/java/com/hazelcast/aws/TagFilters.java]: I have moved this functionality back to  `AwsEc2Api` as suggested although I found it a little bit harder to test if these tag specific filters are built correctly. I'll leave it for you to judge and if needed I can still do some adjustments.

File: src/main/java/com/hazelcast/aws/AwsRequestUtils.java (hunk 4 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsRequestUtils.java:88]: thanks for the explanation 👍

File: src/main/java/com/hazelcast/aws/AwsEc2Api.java [hunk 7]
```diff
@@ -92,6 +86,27 @@
...
+    private void addTagFilter(Filter filter, Tag tag) {
+        if (isNotEmpty(tag.getKey()) && isNotEmpty(tag.getValue())) {
+            filter.add("tag:" + tag.getKey(), tag.getValue());
+            return;
+        }
+        if (isNotEmpty(tag.getKey())) {
+            filter.add("tag-key", tag.getKey());
+            return;
+        }
+        if (isNotEmpty(tag.getValue())) {
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/AwsEc2Api.java:105]: Is this a valid condition? Can you have a value without a key?

File: src/main/java/com/hazelcast/aws/AwsEc2Api.java [hunk 8]
```diff
@@ -92,6 +86,27 @@
...
+            filter.add("tag-value", tag.getValue());
+        }
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/AwsEc2Api.java:107]:
```suggestion
        if (isNotEmpty(tag.getKey()) && isNotEmpty(tag.getValue())) {
            filter.add("tag:" + tag.getKey(), tag.getValue());
        } else if (isNotEmpty(tag.getKey())) {
            filter.add("tag-key", tag.getKey());
        } else if (isNotEmpty(tag.getValue())) {
            filter.add("tag-value", tag.getValue());
        }
```

File: src/main/java/com/hazelcast/aws/CollectionUtils.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package com.hazelcast.aws;
+
+import java.util.Collection;
+
+class CollectionUtils {
+    private CollectionUtils() {
+    }
+
+    static boolean isNotEmpty(Collection<?> collection) {
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/CollectionUtils.java:9]: You use it only in `AwsConfig`, so I think you can just move this method there. No need for a separate utils class.

File: src/main/java/com/hazelcast/aws/Tag.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package com.hazelcast.aws;
+
+class Tag {
```
Comment (leszko) [inline src/main/java/com/hazelcast/aws/Tag.java:3]: Please add some javadoc.

**Review by leszko — CHANGES_REQUESTED** [review]
Added my comments. Apart from this, I think we should fail-fast if the number of tag values is longer than tag keys. That's a wrong configuration since AFAIK you cannot have a tag with value, (but without a key). Could you double-check it?

**Comment by ajurasz** [comment]
In terms of AWS Tags, we cannot have a tag value without specifying its key - it's not a valid tag. 

But please notice that in our case `tag-key` and `tag-value` are representing filtering values - this is what we have got in our javadoc:

> Tag key as specified in the EC2 console. It is used to narrow the members returned by the discovery mechanism.
> Tag value as specified in the EC2 console. It is used to narrow the members returned by the discovery mechanism. 

To keep backwards compatibility I left this functionality unchanged since I can imagine that it can be a valid use case to search for hazelcast members using just `tag-value` and someone could be using it already.

---

What I found a little bit confusing at the time when I was analyzing this task, was that AWS documentation does not specify `tag-value` as one of the available filters options, please see: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html#API_DescribeInstances_RequestParameters

Above is also valid for AWS CLI documentation - it does not list `tag-value` as a filter option, please see: https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html#options

The confusing part is in examples, please see: https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html#examples (point 3, "To describe instances based on tags")

---

Since I know that filtering by `tag-value` works (see https://github.com/hazelcast/hazelcast-aws/issues/202#issuecomment-767117370) I decided to leave it and don't break backwards compatibility. But if you think we should drop the functionality I can "fail-fast" as soon as I recognize tag with value only.

**Comment by leszko** [comment]
> In terms of AWS Tags, we cannot have a tag value without specifying its key - it's not a valid tag.
> 
> But please notice that in our case `tag-key` and `tag-value` are representing filtering values - this is what we have got in our javadoc:
> 
> > Tag key as specified in the EC2 console. It is used to narrow the members returned by the discovery mechanism.
> > Tag value as specified in the EC2 console. It is used to narrow the members returned by the discovery mechanism.

Yes, you're right. I didn't think about this case. Then, your implementation is correct. Please update the other minor comments and then your PR is ready to merge :)

**Review by leszko — APPROVED** [review]
Thanks for all the changes (and for the PR) @ajurasz LGTM 👍
